job-runner/chain/reaper.go

```diff
@@ -0,0 +1,589 @@
+			return
+		}
+
+		// If the failed job was stopped or is not part of a retryable
+		// sequence, then ignore subsequent jobs. For example, if job B of
+		// A -> B -> C  fails, then C is not run.
+		jLogger.Warn("job did not complete successfully.")
+		return
+	}
+	jLogger.Infof("job completed successfully.")
+
+	// Copy job data from this job to all child jobs. When a job has multiple
+	// parent jobs, it will get jobData copied from all of its parent jobs in
+	// the order that they finish. Therefore, if multiple parent jobs try to set
+	// the same jobData field, it is undetermined which parent job the job will
+	// receive the final copy of that jobData field from.
+	for _, nextJob := range r.chain.NextJobs(job.Id) {
+		nextJLogger := jLogger.WithFields(log.Fields{"next_job_id": nextJob.Id})
+
+		for k, v := range job.Data {
+			nextJob.Data[k] = v
+		}
+		r.chainRepo.Set(r.chain)
+
+		// Enqueue any runnable child jobs.
+		if !r.chain.IsRunnable(nextJob.Id) {
+			nextJLogger.Infof("next job is not runnable - not enqueuing it")
+			continue
+		}
+		nextJLogger.Infof("next job is runnable - enqueuing it")
+		if r.chain.IsSequenceStartJob(nextJob.Id) {
+			// Starting a sequence, so increment sequence try count.
+			r.chain.IncrementSequenceTries(nextJob.Id)
+			seqLogger := r.logger.WithFields(log.Fields{"sequence_id": job.SequenceId, "sequence_try": r.chain.SequenceTries(job.Id)})
+			seqLogger.Info("starting try of sequence")
+		}
+		r.send(nextJob)
+	}
+}
+
+// Send a job to runJobChan. This is very simple, but using this method
+// makes sure every job sent to runJobChan is also added to r.running.
+func (r *RunningChainReaper) send(job proto.Job) {
+	r.running[job.Id] = struct{}{}
+	r.runJobChan <- job
+}
+
+// Finalize determines the final state of the chain and sends it to the Request
+// Manager.
+func (r *RunningChainReaper) Finalize() {
+	_, complete := r.chain.IsDoneRunning()
+	if complete {
+		r.logger.Infof("chain is done, all jobs finished successfully")
+		r.chain.SetState(proto.STATE_COMPLETE)
+	} else {
+		r.logger.Warn("chain is done, some jobs failed")
+		r.chain.SetState(proto.STATE_FAIL)
+	}
+	r.chainRepo.Set(r.chain)
+	if r.sendFinalState() {
+		r.chainRepo.Remove(r.chain.RequestId())
+	}
+}
+
+// -------------------------------------------------------------------------- //
+
+// Job Reaper for chains that are being suspended (stopped to be resumed later).
+type SuspendedChainReaper struct {
+	reaper
+	stopTimeout time.Duration
+}
+
+// Run reaps jobs when they finish running. For each job reaped, if it's...
+// - completed: prepare subsequent jobs (copy jobData).
+// - failed:    prepare a sequence retry.
+// - stopped:   do nothing (job will be retried when chain is resumed).
+func (r *SuspendedChainReaper) Run() {
+	// Wait at most 10 seconds for all running jobs to finish.
+	timeoutChan := make(chan struct{})
+	go func() {
+		time.Sleep(r.stopTimeout)
+		close(timeoutChan)
```

Comment:
Oh, I see what it's doing. Instead of this block, you want `case <-time.After(r.stopTimeout):` instead of `case <-timeoutChan:`.  Also, a comment somewhere that explains that this (and the Stopped) reaper only run for a short time to give in-flight jobs time to finish and be reaped. The traverser doesn't explicitly stop this reaper; it needs to stop itself.

Explained that way, I think it makes more sense for caller (traverser) to wait and stop. This internal timeout is kind of magical. Reaper.Run() should run until down and return, or until its Stop method is called. So traverser should swap in this reaper, run it and either wait for it to return or timeout and call its Stop method.